tfep/io/tmbarlog.py

```diff
@@ -0,0 +1,718 @@
+        data_loader : torch.utils.data.DataLoader, optional
+            The data loader used for training wrapping a :class:``tfep.io.dataset.TrajectoryDataset``.
+            This must be passed when a new logger is created as it is used to
+            determine epoch, batch, and trajectory dimensions. If ``save_dir_path``
+            points to an existing logger, then this is ignored.
+        train_subdir_name : str, optional
+            The name of the subdirectory where the training data is stored.
+        eval_subdir_name : str, optional
+            The name of the subdirectory where the evaluation data is stored.
+        n_states : int, optional
+            The number of states in the system. Default is 2.
+        state_names : List[str], optional
+            Names for each state. If None, defaults to ['state_0', 'state_1', ...].
+
+        """
+        self._save_dir_path = os.path.realpath(save_dir_path)
+        self._train_dir_path = os.path.join(save_dir_path, train_subdir_name)
+        self._eval_dir_path = os.path.join(save_dir_path, eval_subdir_name)
+        
+        # Initialize state parameters
+        self._n_states = n_states
+        self._state_names = state_names if state_names is not None else [f'state_{i}' for i in range(n_states)]
+        
+        # Generate all possible state mappings
+        self._state_mappings = []
+        for i in range(n_states):
+            for j in range(n_states):
+                if i != j:  # Exclude self-mappings
+                    self._state_mappings.append((i, j))
+        
+        # This keep track of the currently in-memory training/evaluation data.
+        self._loaded_train_idx = None
+        self._loaded_train_data = None  # Dict[name, Tensor].
+        self._loaded_eval_idx = None
+        self._loaded_eval_data = None  # Dict[name, Tensor].
+
+        # Determine whether this is a new logger or we are resuming. The
+        # metadata file is the last file that is created in __init__().
+        metadata_file_path = os.path.join(save_dir_path, self.METADATA_FILE_NAME)
+        resume = os.path.isfile(metadata_file_path)
+
+        # Load metadata.
+        if resume:
+            self._metadata_from_file(metadata_file_path)
+        elif data_loader is None:
+            raise ValueError("When creating a new logger, 'data_loader' must be passed.")
+        else:
+            self._metadata_from_data(data_loader)
+
+        # Create directory structure.
+        os.makedirs(save_dir_path, exist_ok=True)
+        for dir_path in [self._train_dir_path, self._eval_dir_path]:
+            os.makedirs(dir_path, exist_ok=True)
+
+        # Save metadata. This must be the very last thing to do in __init__ for
+        # resuming to work. No need to write on disk if file is already there.
+        if not resume:
+            self._save_metadata(metadata_file_path)
+
+    @property
+    def batch_size(self):
```

Comment:
I think a lot of this is redundant. Since you're inheriting from `TFEPLogger`, can you remove the bits of codes that are identically implemented in there?